source/agora/common/Deserializer.d

```diff
@@ -188,35 +193,120 @@ public void deserializePart (ref ubyte record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(uint*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
-public void deserializePart (ref ulong record, scope DeserializeDg dg)
+public void deserializePart (ref ulong record, scope DeserializeDg dg, CompactMode compact = CompactMode.Yes)
     @trusted
 {
-    record = *cast(ulong*)(dg(record.sizeof).ptr);
+    if (compact == CompactMode.Yes)
+        deserializeVarInt(record, dg);
+    else
+        record = *cast(ulong*)(dg(record.sizeof).ptr);
 }
 
 /// Ditto
 public void deserializePart (ref char[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = cast(char[])dg(length);
 }
 
 /// Ditto
 public void deserializePart (ref ubyte[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = dg(length);
 }
+
+/*******************************************************************************
+
+    Deserialize an integer of variable length using Bitcoin-style encoding
+
+    VarInt Size
+    size_tag is first a ubyte
+    size_tag <= 0xFC(253)  -- 1 byte   ubyte
+    size_tag == 0xFD       -- 3 bytes  (0xFD + ushort)
+    size_tag == 0xFE       -- 5 bytes  (0xFE + uint)
+    size_tag == 0xFF       -- 9 bytes  (0xFF + ulong)
+
+    Params:
+        T = Type of unsigned integer to deserialize
+        num = Value to store the result of deserialization into
+        dg = source of binary data
+
+    See_Also: https://learnmeabitcoin.com/glossary/varint
+*******************************************************************************/
+
+public void deserializeVarInt (T) (ref T num, scope DeserializeDg dg)
+    @safe
+    if (is(T == ushort) || is(T == uint) || is(T == ulong))
+{
+    const ubyte int_size = dg(ubyte.sizeof)[0];
+    assert(int_size >= 0);
+
+    T read (InType)() @trusted
+    {
+        import std.exception;
+        auto value = *cast(InType*)dg(InType.sizeof).ptr;
+        static if (T.max < InType.max)
+            enforce(value <= T.max);
+        return cast(T)value;
+    }
+
+    if (int_size <= 0xFC)
+        num = cast(T)(int_size);
+    else if (int_size == 0xFD)
+        num = read!ushort();
+    else if (int_size == 0xFE)
+        num = read!uint();
+    else if (int_size == 0xFF)
+        num = read!ulong();
+    else
+        assert(0);
+}
+
+/// For varint
+unittest
+{
+    ubyte[] data = [
+        0x00,                           // ulong.init
+        0xFC,                           // ulong(0xFC) == 1 byte
+        0xFD, 0xFD, 0x00,               // ulong(0xFD) == 3 bytes
+        0xFD, 0xFF, 0x00,               // ulong(0xFE) == 3 bytes
+        0xFD, 0xFF, 0xFF,               // ulong(0xFF) == 3 bytes
+        0xFE, 0x00, 0x00, 0x01, 0x00,   // ushort.max  == 5 bytes
```

Comment:
This looks incorrect. Other places have `ushort.max` serialized as `[0xFE, 0xFF, 0xFF]`

Reply:
Modified.